### DISCUSSION TITLE
Should deleteMany delete everything when empty array is provided?

### DISCUSSION THREAD

**Discussion by simon-lammes** [prisma/prisma#5138] [Q&A] [answered]
Hi, I noticed something 'weird'. But before submitting an issue, I wanted to make sure that this is not intended behavior.
The [`deleteMany`](https://www.prisma.io/docs/concepts/components/prisma-client/crud#deletemany) variable is for deleting related entities with specified criteria. You can omit the 'where' property if you want all entities to be deleted. For example, the following update would delete all posts of a user:
```TypeScript
await prisma.user.update({
    where: {
        id: 1
    },
    data: {
        posts: {
            deleteMany: {}
        }
    }
})
```

This makes sense so far. But you get the same result when providing an empty array like so:

```TypeScript
await prisma.user.update({
    where: {
        id: 1
    },
    data: {
        posts: {
            deleteMany: []
        }
    }
})
```

I think that this is somewhat confusing. With an empty object, I am clearly saying: 'Delete all'. With an empty array, I am saying: 'Delete zero objects'. I can explain why I am coming to this conclusion: I used the 'array approach' for deleting all entities removed by the user. It looked like so:

```TypeScript
deleteMany: removedLabels.map(label => {
  return {
    exerciseId: id,
    labelId: label.id
  };
})
```

This approach works fine when the user removes one or more labels. But when the user removes 0 labels, all associated labels are deleted. **The problem I see is that you would not intuitively expect this behavior when you look at the code.** Therefore, I think one might think about how the API could be improved. I do not yet have concrete suggestions because I first wanted to know whether you guys agree that this aspect of the API is currently suboptimal.

To sum up, I had to change my code like so:

```TypeScript
deleteMany: removeLabels.length > 0 ? removeLabels.map(label => {
  return {
    exerciseId: id,
    labelId: label.id
  };
}) : undefined
```

I can live with that but I do not think it is perfect.

**Comment by ryands17** [comment, marked as answer]
Hey @simon-lammes 👋 

The solution that you have suggested above would be the way to go for now as the API works in a way that if the given value is `undefined`, then everything would match and deleted. So passing `{}` or `[]` has the same effect.

The same goes for API's like `findMany` and `updateMany` as `undefined` is inferred by JavaScript you don't pass anything and that matches all the filters.

So as per JavaScript's working where nothing is passed, it would be correct in that case.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
